### PR TITLE
Fix AppCondition cache

### PR DESCRIPTION
**Pull request by francescaleung**
## Description

`AppCondition` is an eager cache for errors + warnings & isn't environment-aware. This patch changes it to a lazy cache that _is_ environment-aware. When something on the filesystem changes, the whole cache is thrown out, which is erring on the side of refreshing the cache very often--but because it's now a lazy cache, I don't think this is too problematic.

This patch also removes the `resources` -> `state` dependency + the unused `UnattachedResource` struct.

Resolves DTS-379.

## Test plan

Started a studio server, navigated to a task page that used resources that existed in the fallback environment. Task page now does not indicate that resources are missing.

### CODE REVIEW

File: pkg/server/state/state.go [hunk 1]
```diff
@@ -143,6 +146,73 @@ func (s *State) GetEnv(ctx context.Context, envSlug string) (libapi.Env, error)
...
+		}
+		if len(missingResources) > 0 {
+			sort.Strings(missingResources)
+			result.Errors = append(result.Errors, dev_errors.AppError{
+				Level:   dev_errors.LevelWarning,
+				AppName: taskConfig.Def.GetName(),
+				AppKind: "task",
+				Reason:  fmt.Sprintf("Attached resource: %v not found in dev config file or remotely.", missingResources),
+			})
+		}
```
Comment (colinking) [inline pkg/server/state/state.go:197]: Unrelated to your change, but the array brackets in this error message have bothered me before. It seems straightforward to create one error per missing resource. What do you think?

```suggestion
		for _, resource := range missingResources {
			result.Errors = append(result.Errors, dev_errors.AppError{
				Level:   dev_errors.LevelWarning,
				AppName: taskConfig.Def.GetName(),
				AppKind: "task",
				Reason:  fmt.Sprintf("Attached resource %q not found in dev config file or remotely.", resource),
			})
		}
```

File: pkg/server/server.go [hunk 2]
```diff
@@ -411,36 +375,30 @@ type DiscoverOpts struct {
...
+		return nil, errors.Wrap(err, "validating task")
 	}
 	if opts.OverwriteAll {
 		// clear existing tasks, task errors, and views
 		s.state.TaskConfigs.ReplaceItems(map[string]discover.TaskConfig{})
-		s.state.AppCondition.ReplaceItems(map[string]state.AppCondition{})
 		s.state.ViewConfigs.ReplaceItems(map[string]discover.ViewConfig{})
 	}
-	now := time.Now()
+	// Always throw out all AppConditions
```
Comment (colinking) [inline pkg/server/server.go:388]: Adjusted wording to clarify it's a cache.

```suggestion
	// Always invalidate the AppConditions cache.
```

File: pkg/server/server.go [hunk 3]
```diff
@@ -411,36 +375,30 @@ type DiscoverOpts struct {
...
+	s.state.AppCondition.ReplaceItems(map[string]state.AppCondition{})
 	for _, cfg := range opts.Tasks {
-		if _, isUnsupported := warnings.UnsupportedApps[cfg.Def.GetSlug()]; !isUnsupported {
+		if _, isUnsupported := unsupportedApps[cfg.Def.GetSlug()]; !isUnsupported {
 			s.state.TaskConfigs.Add(cfg.Def.GetSlug(), cfg)
```
Comment (colinking) [inline pkg/server/server.go:392]: Unrelated to your change, but it seems to me that overwriting should be done atomically? Could we build up the `map[string]discover.TaskConfig` and then do a call to `s.state.TaskConfigs.ReplaceItems(...)` at the end?

File: pkg/server/server.go [hunk 4]
```diff
@@ -242,42 +237,11 @@ func ValidateTasks(ctx context.Context, resourcesWithEnv map[string]env.Resource
...
-					AppKind: apidev.EntityKindTask,
-					Reason:  fmt.Sprintf("Attached resource: %v not found in dev config file or remotely.", missingResources),
-				},
-			}
-		}
 	}
 
-	return dev_errors.RegistrationWarnings{
-		UnsupportedApps:     unsupportedApps,
-		UnattachedResources: unattachedResources,
```
Comment (colinking) [inline pkg/server/server.go:278]: So this was just never used anywhere, right?

**Review by colinking — APPROVED** [review]
Nice. Thanks for debugging!

File: pkg/server/server.go (hunk 4 above)
Comment (francescaleung) [reply pkg/server/server.go:278]: yea, nothing was reading this. Didn't dig into the git history to figure out when/where this had previously got used.

**Pull request merged**